Added models and generators for atmosphere unit

PatyaKM opened the pull request.

PatyaKM committed:
feat: add temperature stabilization unit architecture

PatyaKM committed:
feat: add device architecture

PatyaKM committed:
feat: add humidity management unit architecture

PatyaKM committed:
feat: add pressure balancing unit architecture

PatyaKM committed:
feat: add pressure stabilization unit architecture

PatyaKM committed:
Merge remote-tracking branch 'refs/remotes/origin/11-microservices-refactoring' into 11-microservices-refactoring

PatyaKM committed:
feat: added CPU usage restriction for Docker

PatyaKM committed:
fix: rename indexes

PatyaKM committed:
fix: rename indexes

PatyaKM committed:
fix: rename indexes

PatyaKM committed:
fix: rename indexes
feat: add speed attribute, start method

PatyaKM committed:
fix: rename indexes
feat: add HumidityManager.is_on attribute

PatyaKM committed:
feat: add AtmosphereUnit model

PatyaKM committed:
Merge branch 'andrwnv:11-microservices-refactoring' into 11-microservices-refactoring

PatyaKM committed:
build: removed unused fields

PatyaKM committed:
refactor: removed unused properties

PatyaKM committed:
refactor: replace device.py in domain

PatyaKM committed:
feat: add atmosphere unit model

PatyaKM committed:
feat: add generator for pressure stabilizer

PatyaKM committed:
feat: add generator for temperature stabilizer

PatyaKM committed:
feat: add generator for humidity manager

PatyaKM committed:
feat: add interface for generators

PatyaKM committed:
feat: add generator for atmosphere unit

andrwnv commented on atmosphere_unit_service/service/domain/device.py:
```diff
@@ -0,0 +1,33 @@
+        self.pressure_stabilizer.is_on 
+        or self.temperature_stabilizer.is_on
+        or self.humidity_manager.is_on
+        ):
```
Вся конструкция может быть упрощена

```python
def start(self) -> None:
    self.is_on = self.pressure_stabilizer.is_on or \ 
        self.temperature_stabilizer.is_on or \
        self.humidity_manager.is_on
```

PatyaKM replied:
Поправил

andrwnv commented on atmosphere_unit_service/service/domain/model.py:
```diff
@@ -0,0 +1,21 @@
+            "sector 12": Device("sector 12"),
+            "sector 13": Device("sector 13"),
+            "sector 14": Device("sector 14"),
+            "sector 15": Device("sector 15"),
```
**Названия секторов:**

1. `Command bridge`
2. `Engineering compartment`
3. `Crew quarters`
4. `Medical bay`
5. `Corridors and passageways`
6. `Cafeteria`
7. `Research laboratory`
8. `Alpha storage area`
9. `Sigma storage area`
10. `Library facilities`
11. `Recreation zones`
12. `Gardens area`
13. `Workbenches`
14. `Rest chambers`
15. `Lounge area`

Также данная конструкция может быть автоматизирована

Для пример:
```python
IDS = [ 1, 2, 3, 4, 5 ]
_map = { f'key-{ix}': ix for ix in IDS } 
print(_map)
```

PatyaKM replied:
Названия добавил, автоматизацию не трогал.

andrwnv commented on atmosphere_unit_service/service/generators/humidity_manager/humidity_manager_generator.py:
```diff
@@ -0,0 +1,47 @@
+                    self.GenerationParams.SIGMA,
+                )
+        else:
+            pass
```
1. Тут должно быть обращение к полям модели, а не сигнатуре класса
2. Зачем `else`?

PatyaKM replied:
Else лишний убрал, обращение поправил

andrwnv commented on atmosphere_unit_service/service/generators/humidity_manager/humidity_manager_generator.py:
```diff
@@ -0,0 +1,47 @@
+        self.__model.alarm = False
+
+        if self.__model.durability <= 27:
+             self.__model.alarm = True
```
аналогичная история как и с методом `start` в классе `Device` - может быть упрощено

PatyaKM replied:
Упростил

andrwnv commented on atmosphere_unit_service/service/generators/pressure_stabilizer/pressure_stabilizer_generator.py:
```diff
@@ -0,0 +1,68 @@
+                self.GenerationParams.N2_VOLUME,
+                self.GenerationParams.SIGMA
+            )
+            self.__model.other_volume = 100-sum(
+                self.__model.CO2_volume,
+                self.__model.O2_volume,
+                self.__model.N2_volume
+            )
```
Потенциально, при некоторых значения Гауссовского распределения может стать меньше 0

PatyaKM replied:
Распределение поправил

andrwnv commented on atmosphere_unit_service/service/generators/pressure_stabilizer/pressure_stabilizer_generator.py:
```diff
@@ -0,0 +1,68 @@
+        else:
+            pass
+        
+        if self.__model.CO2_volume <= 3.0:
+            self.__model.is_suitable = True
+        
+        self.__model.alarm = False
+
+        if self.__model.durability <= 20:
+            self.__model.alarm = True
```
Снова лишние `if`-ы

PatyaKM replied:
От if-ов избавился, упростил

andrwnv commented on atmosphere_unit_service/service/generators/temperature_stabilizer/temperature_stabilizer_generator.py:
```diff
@@ -0,0 +1,45 @@
+        self.__model = model
+        self.__name = name
+    
+    def generate_properties(self) -> None:
+        if self.__model.is_on:
+            TemperatureStabilizer.temperature = random.gauss(
+                self.GenerationParams.TEMPERATURE,
+                self.GenerationParams.SIGMA
+            )
+        else:
+            pass
+
+        self.__model.alarm = False
+
+        if self.__model.durability <= 50:
+            self.__model.alarm = True
```
Полностью аналогично предшествующим тредам

PatyaKM replied:
Else и if убрал, обращение поправил, упростил

andrwnv commented on atmosphere_unit_service/service/domain/humidity_manager/humidity_manager.py:
```diff
@@ -0,0 +1,21 @@
+    alarm: bool
+    durability: float
+
+    sector: str
```
Никогда не заполняется. Зачем оно тут?

PatyaKM replied:
Лишнее убрал

andrwnv commented on atmosphere_unit_service/service/generators/humidity_manager/humidity_manager_generator.py:
```diff
@@ -0,0 +1,47 @@
+        self.__model = model
+        self.__name = name
+
+    def generate_properties(self) -> None:
```
Отсутствует часть логики генерации описанная в тз 

1. Если влажность ниже 50 процентов, то есть (50 - Х) вероятность, что **прочность сектора упадёт на 1 процент** из-за повреждения приборов
2. Если влажность выше 70 процентов, есть (Х - 70) вероятность, что **прочность сектора упадёт на 5 процентов**

Где X - текущая влажность

PatyaKM replied:
Добавил

andrwnv commented on atmosphere_unit_service/service/generators/temperature_stabilizer/temperature_stabilizer_generator.py:
```diff
@@ -0,0 +1,45 @@
+        self.__model = model
+        self.__name = name
+    
+    def generate_properties(self) -> None:
```
Отсутствует логика генерации:
1. Если сектор находится на солнечной стороне, при отказе BT-0010, **температура поднимается до 120**. Если сектор находится на противоположной солнечной стороне, **температура падает до -160**.

PatyaKM replied:
Добавил

andrwnv commented on atmosphere_unit_service/service/generators/pressure_stabilizer/pressure_stabilizer_generator.py:
```diff
@@ -0,0 +1,68 @@
+        self.__model = model
+        self.__name = name
+    
+    def generate_properties(self) -> None:
```
Отсутствует логика генерации:

1. Если внешняя целостность сектора нарушена (прочность сектора меньше или равна 30), атмосферное давление начинает падать со скоростью прямо пропорционально повреждениям сектора. Скорость падения высчитывается по формуле 0.1 * (100/x) атмосферы в секунду, где x - прочность сектора.
2. Если внешняя целостность сектора не нарушена, атмосферное давление падает со скоростью, высчитываемой из п. 1 и домноженной на коэффициент 0.01. При этом уровень С02 растёт со скоростью 1% в минуту. При достижении 5% работа подконтрольного сектора приостанавливается.
3. Если блок в рабочем состоянии, давление **стабилизируется со скоростью 50 мм рт. ст. в секунду**

PatyaKM replied:
Добавил

andrwnv commented on atmosphere_unit_service/service/domain/humidity_manager/humidity_manager.py:
```diff
@@ -0,0 +1,21 @@
+
+    sector: str
+    humidity: float
+    
```
Добавить недостающие newline в конце файлов, это упрощает работу с последующими `diff`-ами

PatyaKM replied:
Где недоставало - добавил

andrwnv reviewed (commented):
В целом неплохо, но не увидел файлов для запуска сервиса, откуда появляется вопрос - как проверялась работоспособность сервиса?

PatyaKM committed:
refactor: add names of sectors

PatyaKM committed:
refactor: code simplification

PatyaKM committed:
refactor: remove unused params

PatyaKM committed:
refactor: add new params

PatyaKM committed:
refactor: add new params

PatyaKM committed:
refactor: code simplification
feat: add generation logic

PatyaKM committed:
refactor: code simplification
feat: add generation logic
fix: correction of values for the Gauss's distribution

PatyaKM committed:
refactor: code simplification
feat: add generation logic
fix: fix Gauss's distribution

PatyaKM committed:
refactor: add missing newlines

PatyaKM commented:
Работоспособность ещё не проверялась. Файлы запуска в процессе написания

andrwnv commented on atmosphere_unit_service/service/generators/humidity_manager/humidity_manager_generator.py:
```diff
@@ -0,0 +1,53 @@
+
+        self.__model.alarm = self.__model.durability <= 27
+
+        if self.__model.humidity < 50:
+            self.GenerationParams.CHANCE = 50 - self.__model.humidity
+            if random(0,100,1) <= self.GenerationParams.CHANCE:
+                self.__model.durability -= 1
+        
+        if self.__model.humidity > 70:
+            self.GenerationParams.CHANCE = self.__model.humidity - 70
+            if random(0,100,1) <= self.GenerationParams.CHANCE:
+                self.__model.durability -= 5
```
Переопределение константных значений

PatyaKM replied:
Так логика генерации предполагает изменение прочности. Как её обозначить, не прибегая к переопределению значений?

andrwnv replied:
Ты тут изначально не совсем прав, ты вынес константой не то что нужно было. 

Шанс влияющий на выпадения события рассчитывается исходя из текущего показателя элемента, т.е. это должно было выглядеть +- как-то так

```python3
class DefaultGenerationStrategy(IGenerationStrategy):
    @dataclass
    class GenerationParams:
        HUMIDITY = 60.0
        SIGMA = 10.0
        
        CRIT_HIGH_HUMIDITY_LEVEL = 70.0
        CRIT_LOW_HUMIDITY_LEVEL = 50.0
        
    # ... def _init_(...): ...
    
    def generate_properties(self) -> None:
        # ...
        if self.__model.humidity < self.GenerationParams.CRIT_LOW_HUMIDITY_LEVEL :
            chance = abs(self.GenerationParams.CRIT_LOW_HUMIDITY_LEVEL - self.__model.humidity)
            if random(0,100,1) <= chance:
                self.__model.durability -= 1

        if self.__model.humidity > self.GenerationParams.CRIT_HIGH_HUMIDITY_LEVEL :
             chance = abs(self.GenerationParams.CRIT_HIGH_HUMIDITY_LEVEL - self.__model.humidity)
             if random(0,100,1) <= chance:
                self.__model.durability -= 5
```

PatyaKM committed:
feat: add files and lines for starting service

andrwnv commented on protos/atmosphere-unit/atmosphere_unit.proto:
```diff
@@ -0,0 +1,66 @@
+package atmosphere_unit;
+
+service AtmosphereUnitService {
+    rpc GetPressureStabilizer(StateRequest) return (PressureStabilizer) {}
+    rpc GetTemperatureStabilizer(StateRequest) return (TemperatureStabilizer) {}
+    rpc GetHumidityManager(StateRequest) return (HumidityManager) {}
```
В `grpc` нет ключевого слова `return`, есть `returns`

```suggestion
    rpc GetPressureStabilizer(StateRequest) returns (PressureStabilizer) {}
    rpc GetTemperatureStabilizer(StateRequest) returns (TemperatureStabilizer) {}
    rpc GetHumidityManager(StateRequest) returns (HumidityManager) {}
```

andrwnv commented on atmosphere_unit_service/MakeFile:
`MakeFile` -> `Makefile`

Иначе утилита `make` не может найти этот файл и использовать его

andrwnv commented on atmosphere_unit_service/service/__init__.py:
Тут должен быть "экспорт" команды `run`, иначе его будет нельзя найти в `main.py`

```python3
from .service import run
```

andrwnv commented on atmosphere_unit_service/service/config.py:
```diff
@@ -0,0 +1,74 @@
+class Config:
+
+    def __init__(self) -> None:
+        self.self._props = SelfProps(
```
Сломанная конструкция

```suggestion
        self.self_props = SelfProps(
```

andrwnv commented on atmosphere_unit_service/service/service.py:
```diff
@@ -0,0 +1,51 @@
+from amqp.amqp_consumer import RmqConsumer
+from amqp.amqp_publisher import RmqPublisher
+
+from service.generator.generator import DomainGenerator
```
Не валидный импорт

```suggestion
from service.generators.generator import DomainGenerator
```

andrwnv commented on atmosphere_unit_service/service/api/grpc/energy_service.py:
`energy_service`?

andrwnv commented on atmosphere_unit_service/service/api/grpc/energy_service.py:
```diff
@@ -0,0 +1,107 @@
+import logging
+import grpc
+
+import proto.atmosphere_unit.energy_unit_pb2 as pb2
+import proto.atmosphere_unit.energy_unit_pb2_grpc as pb2_grpc
+
+from service.domain.device import Device
+from service.domain import Model
```
Не валидные импорты

```suggestion
import proto.atmosphere_unit.atmosphere_unit_pb2 as pb2
import proto.atmosphere_unit.atmosphere_unit_pb2_grpc as pb2_grpc

from service.domain.device import Device
from service.domain.model import Model
```

andrwnv commented on atmosphere_unit_service/service/api/grpc/server.py:
```diff
@@ -0,0 +1,33 @@
+import proto.atmosphere_unit.atmosphere_unit_pb2_grpc as pb2_grpc
+
+from service.domain import Model
+
+from service.api.grpc.atmosphere_service import AtmosphereService
```
Не валидные импорты

```suggestion
from service.api.grpc.energy_service import AtmosphereService
from service.domain.model import Model
```

andrwnv commented on atmosphere_unit_service/service/domain/device.py:
```diff
@@ -0,0 +1,29 @@
+
+        self.pressure_stabilizer = PressureStabilizer(name=f"PressureStabilizer-{uuid.uuid4()}")
+        self.temperature_stabilizer = TemperatureStabilizer(name=f"TemperatureStabilizer-{uuid.uuid4()}")
+        self.humidity_manager = HumidityManager(name-f"HumidityManager-{uuid.uuid4}")
```
Сломанная конструкция

```suggestion
        self.humidity_manager = HumidityManager(name=f"HumidityManager-{uuid.uuid4}")
```

andrwnv commented on atmosphere_unit_service/service/domain/pressure_stabilizer/pressure_stabilizer.py:
```diff
@@ -0,0 +1,41 @@
+        self.O2_volume = 20.0
+        self.N2_volume = 78.0
+        self.other_volume = 100-(
+            sum(
+            self.CO2_volume, 
+            self.N2_volume, 
+            self.N2_volume
+            )
```
Функция `sum` так не работает

```suggestion
            sum([
                self.CO2_volume, 
                self.N2_volume, 
                self.N2_volume
            ])
```

andrwnv commented on atmosphere_unit_service/service/generators/generator.py:
```diff
@@ -0,0 +1,70 @@
+
+from service.domain.model import Model
+
+import service.generators.humidity_manager as humidity_manager_generator
+import service.generators.pressure_stabilizer as pressure_stabilizer_generator
+import service.generators.temperature_stabilizer as temperature_stabilizer_generator
+
+from service.generators.generation_strategy import IGenerationStrategy
+
+logger = logging.getLogger(__name__)
+
+class DomainGenerator:
+    @dataclass
+    class GeneratorPack:
+        humidity_manager_generator: IGenerationStrategy
+        pressure_stabilizer_generator: IGenerationStrategy
+        temperature_stabilizer_generator: IGenerationStrategy
+    
+    def __init__(self, model: Model) -> None:
+        self.__generator = {}
+
+        for device_sector, device in model.device.items():
+            self.__generator[device_sector] = DomainGenerator.GeneratorPack(
+                humidity_manager_generator=humidity_manager_generator.DefaultGenerationStrategy(
+                    device.humidity_manager
+                ),
+                pressure_stabilizer_generator=pressure_stabilizer_generator.DefaultGenerationStrategy(
+                    device.pressure_stabilizer
+                ),
+                temperature_stabilizer_generator=temperature_stabilizer_generator.DefaultGenerationStrategy(
+                    device.temperature_stabilizer
+                ),
+            )
```
Тут частично проблема с импортами и их переопределением в `GeneratorPack`

```suggestion
import service.generators.humidity_manager.humidity_manager_generator as humidity_generator
import service.generators.pressure_stabilizer.pressure_stabilizer_generator as pressure_generator
import service.generators.temperature_stabilizer.temperature_stabilizer_generator as temp_generator

from service.generators.generation_strategy import IGenerationStrategy

logger = logging.getLogger(__name__)

class DomainGenerator:
    @dataclass
    class GeneratorPack:
        humidity_manager_generator: IGenerationStrategy
        pressure_stabilizer_generator: IGenerationStrategy
        temperature_stabilizer_generator: IGenerationStrategy

    def __init__(self, model: Model) -> None:
        self.__generator = {}

        for device_sector, device in model.device.items():
            self.__generator[device_sector] = DomainGenerator.GeneratorPack(
                humidity_manager_generator=humidity_generator.DefaultGenerationStrategy(
                    device.humidity_manager
                ),
                pressure_stabilizer_generator=pressure_generator.DefaultGenerationStrategy(
                    device.pressure_stabilizer
                ),
                temperature_stabilizer_generator=temp_generator.DefaultGenerationStrategy(
                    device.temperature_stabilizer
                ),
            )
```

andrwnv commented on protos/atmosphere-unit/atmosphere_unit.proto:
```diff
@@ -0,0 +1,66 @@
+}
+
+enum Device {
+    COMMAND_BRIDGE = 0;
+    ENGINEERING_COMPARTMENT = 1;
+    CREW QUARTTERS = 2;
+    MEDICAL BAY = 3;
+    CORRIDORS_AND_PASSAGEWAYS = 4;
+    CAFETERIA = 5;
+    RESEARCH_LABORATORY = 6;
+    ALPHA_STORAGE_AREA = 7;
+    SIGMA_STORAGE_AREA = 8;
+    LIBRARY_FACILITIES = 9;
+    RECREATION ZONES = 10;
+    GARDES_AREA = 11;
+    WORKBENCHES = 12;
+    REST_CHAMBERS = 13;
+    LOUNGE_AREA = 14;
```
```suggestion
    COMMAND_BRIDGE = 0;
    ENGINEERING_COMPARTMENT = 1;
    CREW_QUARTTERS = 2;
    MEDICAL_BAY = 3;
    CORRIDORS_AND_PASSAGEWAYS = 4;
    CAFETERIA = 5;
    RESEARCH_LABORATORY = 6;
    ALPHA_STORAGE_AREA = 7;
    SIGMA_STORAGE_AREA = 8;
    LIBRARY_FACILITIES = 9;
    RECREATION_ZONES = 10;
    GARDES_AREA = 11;
    WORKBENCHES = 12;
    REST_CHAMBERS = 13;
    LOUNGE_AREA = 14;
```

Тут почему-то остались пробелы

andrwnv reviewed (commented):
И так, это в целом все проблемы, что мешали запуску сервиса. 

Все эти проблемы исправляются исходя ошибок которые пишет python при попытке запуска сервиса

PatyaKM committed:
Merge branch 'andrwnv:11-microservices-refactoring' into 11-microservices-refactoring

PatyaKM committed:
fix: minor fixes in files for start service

PatyaKM committed:
fix: change ports for atmosphere-unit-service
